packages/ra-auth-msal/src/authProvider.ts

```diff
@@ -0,0 +1,154 @@
+import { AuthProvider } from "react-admin";
+
+const defaultLoginRequest = {
+  scopes: ["User.Read"],
+};
+
+const defaultTokenRequest = {
+  scopes: ["User.Read"],
+  forceRefresh: false,
+};
+
+const defaultGetPermissionsFromAccount = async () => {
+  return [];
+};
+
+const defaultGetIdentityFromAccount = async (account: AccountInfo) => {
+  return {
+    ...account,
+    id: account?.localAccountId,
+    fullName: account?.username,
+  };
+};
+
+export type MsalAuthProviderParams = {
+  msalInstance: PublicClientApplication;
+  loginRequest?: RedirectRequest;
+  tokenRequest?: SilentRequest;
+  getPermissionsFromAccount?: (
+    account: AccountInfo
+  ) => ReturnType<AuthProvider["getPermissions"]>;
+  getIdentityFromAccount?: (
+    account: AccountInfo
+  ) => ReturnType<AuthProvider["getIdentity"]>;
+};
+
+/**
+ * Function that returns an authProvider using the Microsoft Authentication Library (MSAL),
+ * ready to be used with react-admin.
+ * 
+ * @param msalInstance - The MSAL Client instance
+ * @param loginRequest - The login request configuration object
+ * @param tokenRequest - The token request configuration object
+ * @param getPermissionsFromAccount - Function allowing to customize how to compute the permissions from the account info
+ * @param getIdentityFromAccount - Function allowing to customize how to compute the identity from the account info
+ * 
+ * @example
+ * ```tsx
+ * import { PublicClientApplication } from "@azure/msal-browser";
+ * import { LoginPage, msalAuthProvider } from "ra-auth-msal";
+ * import { Admin } from "react-admin";
+ * import { BrowserRouter } from "react-router-dom";
+ * import { msalConfig } from "./authConfig";
+ * import { dataProvider } from "./dataProvider";
+ * 
+ * const myMSALObj = new PublicClientApplication(msalConfig);
+ * 
+ * const App = () => {
+ *   const authProvider = msalAuthProvider({
+ *     msalInstance: myMSALObj,
+ *   });
+ *   
+ *   return (
+ *     <BrowserRouter>
+ *       <Admin 
+ *         authProvider={authProvider} 
+ *         dataProvider={dataProvider}
+ *         loginPage={LoginPage}
+ *       >
+ *         ...
+ *      </Admin>
+ *   </BrowserRouter>
+ *  );
+ * };
+ * ```
+ */
+export const msalAuthProvider = ({
+  msalInstance,
+  loginRequest = defaultLoginRequest,
+  tokenRequest = defaultTokenRequest,
+  getPermissionsFromAccount = defaultGetPermissionsFromAccount,
+  getIdentityFromAccount = defaultGetIdentityFromAccount,
+}: MsalAuthProviderParams): AuthProvider => {
+  // We need to set up the redirect handler at a global scope to make sure all redirects are handled,
+  // otherwise the lib can lock up because a redirect is still marked as pending and has not been handled.
+  // Besides, we can call this handler again later and still gather the response because it is cached internally.
+  msalInstance.handleRedirectPromise();
+
+  return {
+    async login() {
+      // should never be called
+      throw Error("Not implemented");
```

Comment:
It would be nice to allow users to choose whether to redirect immediatly to the third party login on checkAuth or to to do it from a user action. See https://github.com/marmelab/ra-auth-auth0/blob/main/packages/ra-auth-auth0/src/authProvider.ts

Reply:
Right !
Besides, I forgot that we added a [demo-fake-api](https://github.com/marmelab/ra-auth-auth0/tree/main/packages/demo-fake-api) package, which will allow me to demonstrate the `httpClient` feature. I'll add this as well.
Back to WIP.

Reply:
Done, back to RFR!